[Navigation] Add finished promise to NavigationTransition

TingPing opened the pull request:
#### 8467cf6fcc1de1cfa291c50575c4ec1b34387ed5
<pre>
[Navigation] Add finished promise to NavigationTransition
<a href="https://bugs.webkit.org/show_bug.cgi?id=277264">https://bugs.webkit.org/show_bug.cgi?id=277264</a>

Reviewed by Alex Christensen.

This creates a promise in NavigationTransition and resolves it per the spec:

<a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#inner-navigate-event-firing-algorithm">https://html.spec.whatwg.org/multipage/nav-history-apis.html#inner-navigate-event-firing-algorithm</a> (Steps 32.4, 33.4.7)

A test case for success and error were added. The error one currently
fails but will work after #31463 lands.

* LayoutTests/http/wpt/navigation-api/transition-promises-expected.txt: Added.
* LayoutTests/http/wpt/navigation-api/transition-promises.html: Added.
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::rejectFinishedPromise):
(WebCore::Navigation::abortOngoingNavigation):
(WebCore::Navigation::innerDispatchNavigateEvent):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/NavigationTransition.cpp:
(WebCore::NavigationTransition::NavigationTransition):
(WebCore::NavigationTransition::resolvePromise):
(WebCore::NavigationTransition::rejectPromise):
(WebCore::NavigationTransition::finished):
* Source/WebCore/page/NavigationTransition.h:

Canonical link: <a href="https://commits.webkit.org/282085@main">https://commits.webkit.org/282085@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7771b16f3c1a37f385efdf425e709fb9cc040a53

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61883 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41237 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14475 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65863 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12428 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64002 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48923 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12766 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49857 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8595 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64952 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38294 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53599 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30688 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34937 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10832 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11359 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56746 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11136 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67591 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5826 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10901 "Found 60 new test failures: fast/css-grid-layout/floating-empty-grids.html fast/css-grid-layout/mozilla/grid-repeat-auto-fill-fit-002.html fast/css-grid-layout/mozilla/grid-repeat-auto-fill-fit-003.html fast/css-grid-layout/mozilla/grid-repeat-auto-fill-fit-004.html fast/css-grid-layout/mozilla/grid-repeat-auto-fill-fit-005-part-1.html fast/css-grid-layout/mozilla/grid-repeat-auto-fill-fit-005-part-2.html fast/css/aspect-ratio-min-height-replaced.html fast/scrolling/mac/scrollbars/overlay-scrollbar-hovered.html http/tests/site-isolation/draw-after-cross-origin-navigation-between-two-remote-frames.html http/wpt/opener/parent-access-child-via-windowproxy.html ... (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57241 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5851 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53545 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57482 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4773 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9344 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37037 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38121 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39217 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37866 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -438,7 +438,7 @@ void Navigation::resolveFinishedPromise(NavigationAPIMethodTracker* apiMethodTra
 
 // https://html.spec.whatwg.org/multipage/nav-history-apis.html#reject-the-finished-promise
-void Navigation::rejectFinishedPromise(NavigationAPIMethodTracker* apiMethodTracker, Exception&& exception, JSC::JSValue exceptionObject)
+void Navigation::rejectFinishedPromise(NavigationAPIMethodTracker* apiMethodTracker, Exception& exception, JSC::JSValue exceptionObject)
```
Can the Exception& be const?

achristensen07 commented on Source/WebCore/page/NavigationTransition.cpp:
```diff
@@ -32,10 +32,32 @@ namespace WebCore {
 
+void NavigationTransition::resolvePromise()
+{
+    m_finished->resolve();
```
This is used a lot without null checks, but its type indicates it can be null.  One of those should change.

TingPing replied:
Made into a `Ref`.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -739,14 +739,23 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
+                if (m_transition) {
+                    m_transition->rejectPromise(exception);
+                    m_transition = nullptr;
+                }
```
nit: I kind of like this syntax a little better:
if (RefPtr transition = std::exchange(m_transition, nullptr))
    transition->rejectPromise(exception);

I don't have a strong reason for preferring that syntax.  It makes fewer lines.  It also detaches the transition before rejecting the promise and your version does after.  I don't think that matters in this case though.  I'm open to either way.

achristensen07 commented on LayoutTests/http/wpt/navigation-api/transition-promises-expected.txt:
```diff
@@ -0,0 +1,5 @@
+CONSOLE MESSAGE: Unhandled Promise Rejection: Error
+
+PASS NavigationTransition finished promise is fulfilled on success
```
Why don't we make an actual WPT test instead of making a WebKit-only test that uses WPT infrastructure?  Are there really 0 WPT tests that cover this?  We should fix that.

TingPing replied:
There are many WPT tests that cover it, however they are very complex tests that require like 8 events happening and we currently fail in another area.

So this just extracts the one event so it has clear coverage until the others are complete.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -700,8 +700,9 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
-        // FIXME: Create finished promise
-        m_transition = NavigationTransition::create(navigationType, *fromNavigationHistoryEntry);
+        auto& domGlobalObject = *jsCast<JSDOMGlobalObject*>(scriptExecutionContext()->globalObject());
+        JSC::JSLockHolder locker(domGlobalObject.vm());
```
Does this need to still be in scope for the updateURLAndHistory call?  I think it only needs to be in scope for the DeferredPromise::create call.  Could you use { and } to add a scope for this?

achristensen07 commented on Source/WebCore/page/NavigationTransition.cpp:
```diff
@@ -32,10 +32,32 @@ namespace WebCore {
+{
+    if (!m_finishedDOMPromise) {
+        auto& promise = *jsCast<JSC::JSPromise*>(m_finished->promise());
+        // We want to keep this object alive as long as we are.
```
I don't think this comment is helpful.

TingPing committed:
[Navigation] Add finished promise to NavigationTransition
https://bugs.webkit.org/show_bug.cgi?id=277264

Reviewed by Alex Christensen.

This creates a promise in NavigationTransition and resolves it per the spec:

https://html.spec.whatwg.org/multipage/nav-history-apis.html#inner-navigate-event-firing-algorithm (Steps 32.4, 33.4.7)

A test case for success and error were added. The error one currently
fails but will work after #31463 lands.

* LayoutTests/http/wpt/navigation-api/transition-promises-expected.txt: Added.
* LayoutTests/http/wpt/navigation-api/transition-promises.html: Added.
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::rejectFinishedPromise):
(WebCore::Navigation::abortOngoingNavigation):
(WebCore::Navigation::innerDispatchNavigateEvent):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/NavigationTransition.cpp:
(WebCore::NavigationTransition::NavigationTransition):
(WebCore::NavigationTransition::resolvePromise):
(WebCore::NavigationTransition::rejectPromise):
(WebCore::NavigationTransition::finished):
* Source/WebCore/page/NavigationTransition.h:

Canonical link: https://commits.webkit.org/282085@main
